[FIX] calendar: public user no attachment

nd-dew opened the pull request:
### [FIX] calendar: public user no attachment
When public user will create an appointment, the attachments
attached to the related mail template will filtered out by
a `_process_attachments_for_post` method (ref.1)
I happens because attachment's `res_model` is not set to `mail.compose.message`.
Instead it's set to `mail.template` also `create_uid` isn't set to public user,
but to the user creating the template). With this commit attachments will
get copied on send with corrected `res_model`.

### [Reproduce]
- Install modules: appointment,website
- Add an attachment A to email templates: (in Settings/Email_Templates):
	- "Appointment Booked"
- On Website/Appointment: book an appointment (from an incognito window)
- BUG: no attachments are added to the emails (Settings/ Technical/ Email/ Emails)
	- no .ics attachment
	- no A attachment

(ref.1)
https://github.com/odoo/odoo/blob/312b0365e26e1578c5995d55f357b8aec14e5a5b/addons/mail/models/mail_thread.py#L2207C1-L2210

opw-[3685871](https://www.odoo.com/web#id=3685871&view_type=form&model=project.task)

nd-dew commented:
Related fix https://github.com/odoo/odoo/pull/160811
Attachments should probably be tied to `calendar.event`

nle-odoo reviewed (approved):
seems to makes sense to me with how `_process_attachments_for_post` works

reth-odoo commented on addons/calendar/tests/test_calendar.py:
```diff
@@ -381,6 +381,16 @@ def _test_emails_has_attachment(self, partners):
 
+        # Check for attachment when public user creates an event ( simulates scheduling appointment from an incognito window)
+        dummy_partner = self.env['res.partner'].create({'name': 'Dummy User', 'email': 'Dummy@user.com'})
+        self.CalendarEvent.with_user(self.env.ref('base.public_user')).sudo().create({
```
user_public?

reth-odoo commented on addons/calendar/tests/test_calendar.py:
```diff
@@ -381,6 +381,16 @@ def _test_emails_has_attachment(self, partners):
         # no more email should be sent
         _test_one_mail_per_attendee(self, partners)
 
+        # Check for attachment when public user creates an event ( simulates scheduling appointment from an incognito window)
```
```suggestion
        # Check for attachment when public user creates an event (simulates scheduling appointment from an incognito window)
```

reth-odoo commented on addons/calendar/tests/test_calendar.py:
```diff
@@ -381,6 +381,16 @@ def _test_emails_has_attachment(self, partners):
+            'start': "2023-10-06 12:00:00",
+            'stop': "2023-10-06 13:00:00",
+        })
+        _test_emails_has_attachment(self, [dummy_partner])
```
Seems like this does the search as sudo so I'm not sure we're actually testing what we want. Does it fail before the fix?

nd-dew replied:
It does fail.... but you got a point there. Cause I've noticed test is not exactly precisely replicating the flow from the event creation by a public user, but afaik it's not because we search as sudo, it's because we create the event as sudo

nd-dew replied:
While the test does fail/pass fine.... but you got a point there. It doesn't cover fully the desired scenario. Cause I've noticed that test is not exactly precisely replicating the flow from the event creation by a public user.
In the real flow the create_uid is set to the public user id=4
While in my test the create_uid is set to the to the creator of the odoo bot, which isn't public.

reth-odoo commented on addons/calendar/models/calendar_attendee.py:
```diff
@@ -114,7 +114,10 @@ def _send_mail_to_attendees(self, mail_template, force_send=False):
 
-                attachment_ids = mail_template.attachment_ids.ids
+                if mail_template.attachment_ids:
+                    attachment_ids = mail_template.attachment_ids.copy({'res_id': 0, 'res_model': 'mail.compose.message'}).ids
```
copy in a loop == :bomb: 

Can you do `copy_data` before the loop and create as many as self in one batch?

nd-dew replied:
not sure how can I do it, can you say more about it?

reth-odoo commented on addons/calendar/models/calendar_attendee.py:
```diff
@@ -114,7 +114,10 @@ def _send_mail_to_attendees(self, mail_template, force_send=False):
 
-                attachment_ids = mail_template.attachment_ids.ids
+                if mail_template.attachment_ids:
+                    attachment_ids = mail_template.attachment_ids.copy({'res_id': 0, 'res_model': 'mail.compose.message'}).ids
```
Also I know why this works/is necessary but it may deserve a comment

reth-odoo commented on addons/calendar/models/calendar_attendee.py:
```diff
@@ -124,9 +127,9 @@ def _send_mail_to_attendees(self, mail_template, force_send=False):
                         'datas': base64.b64encode(ics_file),
                         'description': 'invitation.ics',
                         'mimetype': 'text/calendar',
+                        'name': 'invitation.ics',
```
Not really worth fixing if we're not touching that line tbh
Polutes the blaming process

nd-dew replied:
Actually it got added in the meantime by a different PR

reth-odoo reviewed (commented):
Henlo :fish:

nd-dew commented:
## 1st problem seems solved.
Namely no .ics attachment, seems like this is solved.
- ics is there in the email
- the flow of https://github.com/odoo/odoo/pull/160811 is not broken

Showing both in the video here, where I go over both issues after applying changes from the respective commit, i.e.:
```diff
                   'description': 'invitation.ics',
                   'mimetype': 'text/calendar',
-                  'res_id': event_id,
-                  'res_model': 'calendar.event',
+                  'res_id': 0,
+                  'res_model': 'mail.compose.message',
                   'name': 'invitation.ics',
```
https://github.com/odoo/odoo/assets/33809926/e758a915-f670-41ad-a434-8f6d859949a1

## Let's jump onto issue number 2: no custom attachment

reth-odoo commented on addons/calendar/models/calendar_attendee.py:
```diff
@@ -109,12 +109,36 @@ def _send_mail_to_attendees(self, mail_template, force_send=False):
+            #   attendee2: [attach1_for_attendee2, attach2_for_attendee2]
+            # }
+            number_of_attachments_per_attende = len(mail_template.attachment_ids)
+            attendee_id_attachment_id_map = {}
+            for i, attendee_id in enumerate(self.ids):
+                attendee_attachments_ids = attachment_for_all_attendees_ids[i:i + number_of_attachments_per_attende]
+                attendee_id_attachment_id_map[attendee_id] = attendee_attachments_ids
```
In fact you can still make this with a dict comprehension I think
```
{
    attendee: attendee_attachment_ids[n * template_attachment_count: n * (template_attachment_count + 1)]
    for n, attendee in numerate(attendee)
}
```
It's usually more efficient, I think. And I'd argue more readable.

Would also be fine not to have a dict and to get the attachments in the next loop as they're all the same anyway.

reth-odoo commented on addons/calendar/tests/test_calendar.py:
```diff
@@ -381,6 +381,19 @@ def _test_emails_has_attachment(self, partners):
         # no more email should be sent
         _test_one_mail_per_attendee(self, partners)
 
+        # Check for attachment when public user creates an event ( simulates scheduling appointment from an incognito window)
+        self.new_partner = self.env['res.partner'].create({
+            'name': 'partner_created_on_the_spot_by_the_appointment_form',
+            'email': 'partner_created_on_the_spot_by_the_appointment_form@gmail.com',
+        })
+        self.CalendarEvent.with_user(self.env.ref('base.public_user')).sudo().create({
+            'name': "publicUserEvent",
+            'partner_ids': [(6, False, self.new_partner.ids)],
+            'start': "2023-10-06 12:00:00",
+            'stop': "2023-10-06 13:00:00",
+        })
+        _test_emails_has_attachment(self, [self.new_partner])
```
Would be nice to test with multiple attachments on the template as that's the trickier case

nd-dew commented:
Hey @reth-odoo thanks for having a look, appreciate all comments ;)

In the above push:
- [x] Modified test to check for multiple attachments case
- [x] added multiple partners in the test 
(so for-loop slicing attachments is properly tested, especially since we both made a mistake there :laughing: ) 

---
- [x] handle copy multiple attachments 
- [x] imp comments, names

reth-odoo reviewed (approved):
Seems fine, thanks for the followup and your patience!

nd-dew commented:
Thanks for help with this one @reth-odoo 
Can you r+

reth-odoo commented:
I can't but I've heard of people who can
@tde-banana-odoo Hello :wave:

robodoo commented:
@nd-dew the target branch 'saas-16.3' has been disabled, you may want to close this PR.

nd-dew commented:
Retargetting -> 17.0

robodoo commented:
This PR targets the disabled branch odoo/odoo:saas-16.3, it needs to be retargeted before it can be merged.

tde-banana-odoo commented on addons/calendar/models/calendar_attendee.py:
```diff
@@ -116,12 +116,29 @@ def _send_mail_to_attendees(self, mail_template, force_send=False):
+            template_attachment_count = len(mail_template.attachment_ids)
+            attendee_id_attachment_id_map = {
+                attendee_id: attendee_attachment_ids[i * template_attachment_count: (i + 1) * template_attachment_count]
+                for i, attendee_id in enumerate(self.ids)
```
Can't you make it simpler using ``split_every`` and split ``attendee_attachment_ids`` by number of attendees ?

tde-banana-odoo commented on addons/calendar/models/calendar_attendee.py:
```diff
@@ -109,12 +109,36 @@ def _send_mail_to_attendees(self, mail_template, force_send=False):
+
+            # Prepare attachments values for each attendee (the same values for each attendee)
+            attachments_values = mail_template.attachment_ids.copy_data({'res_id': 0, 'res_model': 'mail.compose.message'})
+            attachments_values *= len(self)
```
Yep... a bit worried for this one, but it allows to avoid copying data numerous times, so...

tde-banana-odoo commented on addons/calendar/tests/test_calendar.py:
```diff
@@ -381,6 +381,30 @@ def _test_emails_has_attachment(self, partners):
         _test_one_mail_per_attendee(self, partners)
 
+        # Check for attachments when public user creates an event ( simulates scheduling appointment from an incognito window)
+        attachment_2 = self.env['ir.attachment'].create({
```
Can't we directly add the second attachment at beginning of test and consider we do all tests in this function with 2 attachments ?

nd-dew replied:
yup, moved it in there

tde-banana-odoo commented on addons/calendar/tests/test_calendar.py:
```diff
@@ -381,6 +381,30 @@ def _test_emails_has_attachment(self, partners):
+
+        self.partner_staff = self.env['res.partner'].create({
+            'name': 'partner_staff',
+            'email': 'partner_staff@gmail.com',
```
Don't use real emails (use @example.com or stuff like that), better safe than sorry :)

tde-banana-odoo reviewed (commented):
Few comments on my side :)

tde-banana-odoo commented:
Zboing @arj-odoo  does it look good for you ?

arj-odoo commented:
> Zboing @arj-odoo does it look good for you ?

Yep, LGTM. Thanks for deeply reviewing it.

nd-dew commented:
@arj-odoo can you r+ ?

tde-banana-odoo commented:
@robodoo r+
